### ISSUE TITLE
Deepspeed sparse attention error

### ISSUE THREAD

**Issue by kobiso** [lucidrains/DALLE-pytorch#38]
Hello!
I've been using `DeepSpeed sparse attention` pretty well, but I got this error since [tag/0.0.59](https://github.com/lucidrains/DALLE-pytorch/releases/tag/0.0.59) release.

```bash
Traceback (most recent call last):
  File "train_DALLE.py", line 300, in <module>
    main()
  File "train_DALLE.py", line 219, in main
    loss = dalle(caption_tokens, images, mask=mask, return_loss=True)
  File "/opt/conda/lib/python3.7/site-packages/torch/nn/modules/module.py", line 727, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/home/shared/workspace/torch_research/text-to-image/dalle-pytorch/models/model_arch.py", line 421, in forward
    out = self.transformer(tokens, mask = mask)
  File "/opt/conda/lib/python3.7/site-packages/torch/nn/modules/module.py", line 727, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/dalle_pytorch/transformer.py", line 106, in forward
    return self.layers(x, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/torch/nn/modules/module.py", line 727, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/dalle_pytorch/reversible.py", line 139, in forward
    x = x + f(x, **f_args)
  File "/opt/conda/lib/python3.7/site-packages/torch/nn/modules/module.py", line 727, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/dalle_pytorch/transformer.py", line 34, in forward
    return self.fn(self.norm(x), **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/torch/nn/modules/module.py", line 727, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/dalle_pytorch/attention.py", line 314, in forward
    if self.noncausal_attn_len:
  File "/opt/conda/lib/python3.7/site-packages/torch/nn/modules/module.py", line 779, in __getattr__
    type(self).__name__, name))
torch.nn.modules.module.ModuleAttributeError: 'SparseAttention' object has no attribute 'noncausal_attn_len'
```

I guess the error occurs because `self.noncausal_attn_len` is removed from `Attention` class in https://github.com/lucidrains/DALLE-pytorch/commit/95ce537dcbe1afb06fed405008afc642233a5199 this update.

I hope `DeepSpeed sparse attention` would continue to be updated, because it reduces model size by half, speeds up 2x, and final performance seems better :)

**Comment by lucidrains** [comment]
@kobiso oh yes, you are correct! fixed in 0.0.60!

**Comment by lucidrains** [comment]
@kobiso have you had any success in training?? how many resnet blocks did you end up using for the discrete VAE?

**Comment by kobiso** [comment]
@lucidrains 
Thank you for the quick response!
Here is my brief experimental results.

### VAE

I did not use resnet blocks for discrete VAE cuz I saw https://github.com/lucidrains/DALLE-pytorch/issues/10#issuecomment-758038919 this issue.
I tried not to spend too much time on VAE, and focused on DALLE instead.
So, I cannot say much about resnet blocks on VAE.

#### VAE model
```python
vae = DiscreteVAE(
    image_size=128,
    num_layers=3,
    num_tokens=2048,
    codebook_dim=256,
    hidden_dim=128,
    temperature=0.9,
).cuda()
```

#### Result
- Dataset: CUB200
- First row: GT
- Second row: soft generation w/ gumbel softmax
- Third row: hard generation w/o gumbel softmax

 
![image](https://user-images.githubusercontent.com/12422441/107869861-37040880-6ed6-11eb-8923-49fb1be3967a.png)

### DALLE on CUB200 (Small dataset)
- Training DALLE on small dataset seems to be okay.

#### Model and training
- Tips
  - DeepSpeed sparse attention is awesome: requires half memory, 2x faster training speed, better performance
  - Learning rate decay helps: personally, i'm using `ReduceLROnPlateau`
  - To speed up training, Automatic Mixed Precision (AMP) is also helpful

```python
        dalle = DALLE(
            dim=256
            vae=vae, 
            num_text_tokens=7800 # this is up to dataset
            text_seq_len=128,  # this is up to dataset
            depth=32
            heads=16
            dim_head=64
            reversible=False,
            attn_types = ('full', 'sparse')
        ).cuda()
```

#### Result
- During training
  - First row: GT
  - Second row: generated images from each text (these are from training set, so could be overfitted)
![image](https://user-images.githubusercontent.com/12422441/107869906-e7720c80-6ed6-11eb-98f2-b3005ab60db8.png)
![image](https://user-images.githubusercontent.com/12422441/107869907-e9d46680-6ed6-11eb-80a0-f03b0caf002b.png)


- Test with random text
  - It seems to generate birds-like images, but i don't think the model actually understand the meaning of each word.
```python
input_text = ["this small white bird has light gray primaries , and a pointed bill",
              "this small white bird has light gray primaries",
              "this small white bird",
              "this bird has a grey crown with a white belly and small black beak",
              "this bird has a grey crown",
              "this bird"
             ]
```
![image](https://user-images.githubusercontent.com/12422441/107869941-3b7cf100-6ed7-11eb-9aaa-57cb6aa7fa7d.png)


### DALLE on COCO (Medium dataset)
- I'm trying to train a DALLE model with COCO dataset, but I encountered NaN problem.
- It is hard to debug because NaN occurs in the middle of the training.

![image](https://user-images.githubusercontent.com/12422441/107870078-d6c29600-6ed8-11eb-82b4-07841fb132a3.png)

- Generated image before NaN occurs

![image](https://user-images.githubusercontent.com/12422441/107870104-1d17f500-6ed9-11eb-88a6-dbc1b02e2ec9.png)
![image](https://user-images.githubusercontent.com/12422441/107870099-14bfba00-6ed9-11eb-9d07-e24df3a33a36.png)

**Comment by lucidrains** [comment]
@kobiso are you training DALL-E with gradient clipping? (Looking at the Naan error)

**Comment by lucidrains** [comment]
@kobiso also wanted to bring to your attention these new types of sparse attention https://github.com/lucidrains/dalle-pytorch#sparse-attention , likely more accurate to what OAI used than the deepspeed sparse attention

**Comment by kobiso** [comment]
@lucidrains 

### NaN problem
I only tried to skip current step when NaN occurs.
This allows to keep training process alive, but I don't think this is a proper solution because NaN occurs a lots after 😓 
I will try gradient clipping and share the result.

![image](https://user-images.githubusercontent.com/12422441/107895851-0bd1f580-6f78-11eb-8ef9-f28e76524fea.png)


### Sparse attention
I tried `attn_types = ('full', 'axial_row', 'axial_col', 'conv_like')` this attention, but the model was too big and slow to train in one gpu.
I guess this should be tried on multi-gpus.
Moreover, loss graph seems weird by having very small loss and large fluctuation occasionally. (Compare it with `DALLE on COCO` loss graph above)
I will try this more and share the result as well :)

![image](https://user-images.githubusercontent.com/12422441/107896251-0d4fed80-6f79-11eb-9ad2-032489a6b6d6.png)

**Comment by lucidrains** [comment]
@kobiso I just found a bug with the sparse conv and sparse axial attention masking :sweat: I forgot there was a <bos> token at the beginning. I think it is fixed now :crossed_fingers: 

ok, sounds good! gradient clipping should help!

**Comment by kobiso** [comment]
@lucidrains 
Awesome! 
I will try again with the sparse attention and gradient clipping :)

**Comment by lucidrains** [comment]
@kobiso if you are in the experimental mood, you can try out adaptive gradient clipping from the recent normalization-free nets paper! https://gist.github.com/lucidrains/0d6560077edac419ab5d3aa29e674d5c

**Comment by lucidrains** [comment]
@kobiso if you are in the experimental mood, you can also try the new adaptive gradient clipping from https://arxiv.org/abs/2102.06171 I believe it should be an even better type of gradient clipping than the naive method before

https://gist.github.com/lucidrains/0d6560077edac419ab5d3aa29e674d5c

**Comment by kobiso** [comment]
@lucidrains 
I was reading the paper as well! I will try the adaptive gradient clipping for sure.
Thanks for sharing!

**Comment by kobiso** [comment]
Quick result share :)

I tried `torch gradient clipping` and `adaptive gradient clipping`, and I got NaN with `adaptive gradient clipping` at the early training steps.
`torch gradient clipping` is still on training, and no NaN occurs until now.
I guess `adaptive gradient clipping` should be inspected more.
I will close this issue and share more results in Discussions tab. :)

- Models
  - Pink: torch gradient clipping
  - Gray: adaptive gradient clipping from https://arxiv.org/abs/2102.06171

![image](https://user-images.githubusercontent.com/12422441/108062996-ee974700-709d-11eb-88df-b16f0847590e.png)
